winch: Implement new trampolines

saulecabrera opened the pull request:
This change is a follow-up to
https://github.com/bytecodealliance/wasmtime/pull/6262, in which the new trampolines, described [here](https://github.com/bytecodealliance/rfcs/blob/main/accepted/tail-calls.md#new-trampolines-and-vmcallercheckedanyfunc-changes), were introduced to Wasmtime.

This change, focuses on the `array-to-wasm`,
`native-to-wasm` and `wasm-to-native` trampolines to restore Winch's working state prior to the introduction of the new trampolines. It's worth noting that the new approach for trampolines make it easier to support the `TypedFunc` API in Winch. Prior to the introduction of the new trampolines, it was not obvious how to approach it.

This change also introduces a pinned register that will hold the `VMContext` pointer, which is loaded in the `*-to-wasm`  trampolines; the `VMContext`  register is a pre-requisite to this change to support the `wasm-to-native` trampolines.

Lastly, with the introduction of the `VMContext` register and the `wasm-to-native` trampolines, this change also introduces support for calling function imports, which is a variation of the already existing calls to locally defined functions.

The other notable piece of this change aside from the trampolines is `winch-codegen`'s dependency on `wasmtime-environ`. Winch is so closely tied to the concepts exposed by the wasmtime crates that it makes sense to tie them together, even though the separation provides some advantages like easier testing in some cases, in the long run, there's probably going to be less need to test Winch in isolation and rather we'd rely more on integration style tests which require all of Wasmtime
pieces anyway (fuzzing, spec tests, etc).

This change doesn't update the  existing implmenetation of `winch_codegen::FuncEnv`, but the intention is to update that part after this change.

<!--
Please make sure you include the following information:

- If this work has been discussed elsewhere, please include a link to that
  conversation. If it was discussed in an issue, just mention "issue #...".

- Explain why this change is needed. If the details are in an issue already,
  this can be brief.

Our development process is documented in the Wasmtime book:
https://docs.wasmtime.dev/contributing-development-process.html

Please ensure all communication follows the code of conduct:
https://github.com/bytecodealliance/wasmtime/blob/main/CODE_OF_CONDUCT.md
-->

alexcrichton commented on crates/winch/src/compiler.rs:
```diff
@@ -92,8 +92,22 @@ impl wasmtime_environ::Compiler for Compiler {
+        let env = FuncEnv::new(&translation.module, translation.get_types(), &self.isa);
+        let buffer = self
+            .isa
+            .compile_trampoline(&wasm_ty, index.as_u32(), &env, TrampolineKind::ArrayToWasm)
```
Not 100% related to this PR per se, but I've found that using all the `FooIndex` types has been quite helpful throughout Cranelift and the existing wasm lowering to ensure that everything is kept straight. For example I saw this and immediately thought "wait that's bad because this is a defined index" and then looked around a bit before I noticed above that the variable was shadowed with the `FuncIndex`. This `index.as_u32()` is fine as-is, but it might be useful to use the `wasmtime-types::*` index types throughout winch in the future (only if you agree of course!)

alexcrichton commented on crates/winch/src/compiler.rs:
```diff
@@ -102,17 +116,46 @@ impl wasmtime_environ::Compiler for Compiler {
+        let env = FuncEnv::new(&translation.module, translation.get_types(), &self.isa);
+        let buffer = self
+            .isa
+            .compile_trampoline(&wasm_ty, 1u32, &env, TrampolineKind::WasmToNative)
```
Instead of manufacturing a dummy `1u32` here could the index of what to call perhaps be a payload of `TrampolineKind`? That way `WasmToNative` would have no payload but the other two would have payloads.

alexcrichton commented on winch/environ/src/lib.rs:
```diff
@@ -31,11 +33,33 @@ impl<'a> winch_codegen::FuncEnv for FuncEnv<'a> {
         }
     }
+
+    fn vmnative_call_host_func_native_call(&self) -> u8 {
+        let pointer_size = self.isa.triple().pointer_width().unwrap().bytes();
+        pointer_size.vmnative_call_host_func_context_func_ref()
+            + pointer_size.vm_func_ref_native_call()
+    }
+
+    fn vmctx_vmfunction_import_wasm_call(&self, index: u32) -> u32 {
+        let func_index = FuncIndex::from_u32(index);
+        self.offsets.vmctx_vmfunction_import_wasm_call(func_index)
+    }
+
+    fn vmctx_vmfunction_import_vmctx(&self, index: u32) -> u32 {
+        let func_index = FuncIndex::from_u32(index);
+        self.offsets.vmctx_vmfunction_import_vmctx(func_index)
+    }
```
Out of curiosity, is there any particular reason to have accessor methods like this as opposed to exposing the `VMOffsets` directly?

alexcrichton commented on winch/codegen/src/isa/x64/mod.rs:
```diff
@@ -120,15 +120,33 @@ impl TargetIsa for X64 {
+        );
+
+        match kind {
+            ArrayToWasm => trampoline.emit_array_to_wasm(ty, func_index)?,
+            NativeToWasm => trampoline.emit_native_to_wasm(ty.clone(), func_index)?,
+            WasmToNative => trampoline.emit_wasm_to_native(&ty)?,
```
minor nit, but here instead of `ty` vs `ty.clone()` vs `&ty` could these all take `ty: &FuncType`?

alexcrichton commented on winch/codegen/src/trampoline.rs:
```diff
@@ -79,14 +86,17 @@ where
+        self.prologue_with_callee_saved();
+        // Move the VM context pointer to the designated, pinned register.
+        self.masm
+            .mov(vmctx_ptr, <A as ABI>::vmctx_reg().into(), OperandSize::S64);
```
Technically this `S64` is actually "whatever the `A`'s pointer size is", right?

alexcrichton commented on winch/codegen/src/trampoline.rs:
```diff
@@ -178,23 +164,196 @@ where
+    }
+
+    fn callee_and_caller_vmctx() -> Vec<ValType> {
+        vec![ValType::I64, ValType::I64]
```
Like above, technically this `I64` should conditionally be `I32` for 32-bit hosts, right? Not that we're going to support those any time soon, but might be good to perhaps abstract ahead of time in case anyone in the future comes to add support if it's not too hard. (if it is a bit onerous I think a `// FIXME: ...` is fine too since we don't have 32-bit support anywhere on the horizon)

alexcrichton reviewed (commented):
Nice!

alexcrichton commented:
Random thought about pinned registers -- while this isn't a problem for winch yet the final set of trampolines are the libcall trampolines which enter the runtime for instructions like `memory.grow`. These trampolines are still handwritten asm and don't come from Cranelift, and they don't handle anything about pinned registers and such at this time. Not to say they couldn't, just something to keep in mind.

saulecabrera commented:
@alexcrichton I've addressed all of your comments; I think this is ready for another pass. I've opted to leave the translation between `ptr_size <> ValType` as a TODO for now. 

Regarding your other comments, I've introduced `wasmtime-environ` as a dependency to `winch-codegen` as discussed during our last meeting, which allows for a direct usage of  `VMOffsets` and `FooIndex` types across trampolines. After this PR lands, I'll update the other pre-existing code that uses `winch_environ::FuncEnv`.

saulecabrera commented:
> I'm not exactly sure why the MIRI check is failing.

I realized that all the other tests are ignoring miri, so I applied the same attribute to Winch's tests.

saulecabrera commented:
Agreed; @cfallin, added you as a reviewer, let me know if you have any thoughts!

cfallin commented:
> Agreed; @cfallin, re-added you as a reviewer, let me know if you have any thoughts!

Yep, I'll take a look now!

cfallin reviewed (approved):
Overall this looks very reasonable -- no objections (and thanks Alex for doing the detailed review here)!
